mesonbuild/optinterpreter.py

```diff
@@ -166,60 +191,86 @@ def func_option(self, args: T.Tuple[str], kwargs: 'FuncOptionArgs') -> None:
             raise OptionException('Option name %s is reserved.' % opt_name)
 
         opt_type = kwargs['type']
-        parser = self.option_types.get(opt_type)
-        if not parser:
-            raise OptionException(f'Unknown type {opt_type}.')
+        parser = self.option_types[opt_type]
         description = kwargs['description'] or opt_name
 
         # Only keep in kwargs arguments that are used by option type's parser
         # because they use @permittedKwargs().
-        known_parser_kwargs = {'value', 'choices', 'yield', 'min', 'max'}
-        parser_kwargs = {k: v for k, v in kwargs.items() if k in known_parser_kwargs and v is not None}
-        opt = parser(description, T.cast('ParserArgs', parser_kwargs))
+        opt = parser(description, kwargs)
```

Comment:
The comment here is no longer accurate after this commit...

But I'm not sure this is the right thing to do. It's now possible to pass e.g. `choices: ...` to a bool option, and the typed kwargs for `option()` permit this, but the typed_kwargs for `boolean_parser()` simply ignore it and then never use it.

By passing only known_parser_kwargs on, we skipped type/description/deprecated as well as anything that was never explicitly passed -- because typed_kwargs filled it in as None.

I think the right thing to do here is to go the other way -- use allow_unknown on the `option()` function, stop type-checking options we don't care about yet, then strip out the 3 that aren't relevant to the suboptions and pass the rest onward.

Reply:
> I think the right thing to do here is to go the other way -- use allow_unknown on the option() function, stop type-checking options we don't care about yet, then strip out the 3 that aren't relevant to the suboptions and pass the rest onward.

+1

Reply:
Yeah, I think that does make more sense. I wen with this approach because it didn't require rewriting the keyword arguments, but you're right, the code is much clearer the other way.